dialogue-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientChannel.java

```diff
@@ -0,0 +1,217 @@
+import com.google.common.base.Strings;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Headers;
+import com.palantir.dialogue.HttpMethod;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.RequestBody;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.UrlBuilder;
+import com.palantir.dialogue.blocking.BlockingChannel;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.RequestBuilder;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.message.BasicHeader;
+
+final class ApacheHttpClientChannel implements BlockingChannel {
+
+    private final CloseableHttpClient client;
+    private final UrlBuilder baseUrl;
+
+    ApacheHttpClientChannel(CloseableHttpClient client, URL baseUrl) {
+        this.client = client;
+        // Sanitize path syntax and strip all irrelevant URL components
+        Preconditions.checkArgument(
+                null == Strings.emptyToNull(baseUrl.getQuery()),
+                "baseUrl query must be empty",
+                UnsafeArg.of("query", baseUrl.getQuery()));
+        Preconditions.checkArgument(
+                null == Strings.emptyToNull(baseUrl.getRef()),
+                "baseUrl ref must be empty",
+                UnsafeArg.of("ref", baseUrl.getRef()));
+        Preconditions.checkArgument(
+                null == Strings.emptyToNull(baseUrl.getUserInfo()), "baseUrl user info must be empty");
+        this.baseUrl = UrlBuilder.from(baseUrl);
+    }
+
+    @Override
+    public Response execute(Endpoint endpoint, Request request) throws IOException {
+        // Create base request given the URL
+        UrlBuilder url = baseUrl.newBuilder();
+        endpoint.renderPath(request.pathParams(), url);
+        request.queryParams().forEach(url::queryParam);
+        URL target = url.build();
+        RequestBuilder builder = buildFor(endpoint, target);
+
+        // Fill headers
+        for (Map.Entry<String, String> header : request.headerParams().entrySet()) {
+            builder.addHeader(header.getKey(), header.getValue());
+        }
+
+        if (request.body().isPresent()) {
+            if (endpoint.httpMethod() == HttpMethod.GET) {
+                throw new SafeIllegalArgumentException("GET endpoints must not have a request body");
+            }
+            if (endpoint.httpMethod() == HttpMethod.DELETE) {
```

Comment:
is this a limitation of the client or are we just enforcing this?

Reply:
Was just enforcing this for consistency, but I've fixed all clients to support delete with a body.